content/technotes/2018-09-09-github-badges.md

```diff
@@ -0,0 +1,395 @@
+                silent = TRUE)
+  if(inherits(readme, "try-error")){
+    return(NULL)
+  }else{
+    
+    badges <- codemetar::extract_badges(readme$download_url)
+    
+    if(nrow(badges)>0){
+      badges$owner <- owner
+      badges$repo <- repo
+    }
+    
+    return(badges)
+  }
+  
+  
+}
+
+.get_badges %>%
+  ratelimitr::limit_rate(ratelimitr::rate(1, 1)) -> get_badges
+
+purrr::map2_df(github_cran$owner, 
+               github_cran$repo,
+              get_badges) %>%
+  readr::write_csv(path = "data/aaall_badges.csv")
+```
+
+Don’t judge me by my filenaming skills. I was *maybe* a bit too
+enthusiastic!
+
+Analyze badges from CRAN packages
+---------------------------------
+
+I wanted to answer several questions about the badges of CRAN packages,
+beyond being just happy to have been able to collect so many of them.
+
+### How many repos have at least one badge?
+
+``` r
+github_cran <- readr::read_csv("data/github_cran_links.csv")
+# the same repo can have been used by several packages!
+badges <- readr::read_csv("data/aaall_badges.csv")
+badges <- dplyr::distinct(badges)
+
+nobadges <- dplyr::anti_join(github_cran, badges,
+                             by = c("owner", "repo"))
+```
+
+There are 1237 packages without any badge (or rather said, without any
+badge that we identified) out of a sample of 3542 packages.
+
+### Among the repos with badges, how many badges?
+
+``` r
+library("magrittr")
+badges %>%
+  dplyr::count(repo, owner,
+               sort = TRUE) -> badges_count
+
+badges_count %>%
+  dplyr::summarise(median = median(n))
+```
+
+    ## # A tibble: 1 x 1
+    ##   median
+    ##    <int>
+    ## 1      4
+
+``` r
+library("ggplot2")
+badges_count %>%
+  ggplot() +
+  geom_histogram(aes(n))+
+  hrbrthemes::theme_ipsum(base_size = 12, 
+                          axis_title_size = 12, 
+                          axis_text_size = 12) +
+  ggtitle("Number of badges per repo",
+          subtitle = "Among repos with at least one badge")
+```
+
+![](/img/blog-images/2018-09-09-github-badges/unnamed-chunk-2-1.png)
+
+The median number of badges is 4, which corresponds to my gut feeling
+that the answer would be “a few”. I have a new question, what’s the repo
+with the most badges?
+
+``` r
+badges_count[1,]
+```
+
+    ## # A tibble: 1 x 3
+    ##   repo     owner               n
+    ##   <chr>    <chr>           <int>
+    ## 1 psycho.R neuropsychology    14
+
+You can browse it at <https://github.com/neuropsychology/psycho.R>.
+Quite a colourful README!
+
+### How many unique badges are there?
+
+For counting types of badges, I’ll use the domain of `image_link`. This
+is an approximation, since e.g. www.r-pkg.org offers several badges.
+
+``` r
+badges %>%
+  dplyr::pull(image_link) %>%
+  purrr::map_df(urltools::url_parse) -> parsed_image_links
+
+parsed_image_links %>%
+  dplyr::pull(domain) %>%
+  unique() %>%
+  sort() -> unique_domains
+
+length(unique_domains)
+```
+
+    ## [1] 85
```

Comment:
wow! I think that's a lot... you might rhetorically ask the user to estimate to themselves how many types of badges you'll find...

Reply:
Good idea. I was expecting more since there are many repos but I need to add that some domains are not badge domains, e.g. ropensci.org

Reply:
ah, right, makes sense.  I wonder if you could filter by img links that are above the first h1 header level?  more work but might show off the power of this approach too...

Reply:
I prefer the image size approach because you could have badges somewhere else 🤔